x/stakeibc/keeper/validator_selection.go

```diff
@@ -4,60 +4,249 @@ import (
+// The unbalanced portion represents delegations from LSM Tokens that are naturally
+//    misaligned with the validator. They must be rebalanced periodically
+// The balance portion represents either native delegations or LSM delegation that have
+//    already been rebalanced. This portion only requires a rebalancing if the validator weights change
+func (k Keeper) RebalanceDelegations(ctx sdk.Context, chainId string, delegationType types.DelegationType, numRebalance uint64) error {
+	// Get the host zone and confirm the delegation account is initialized
+	hostZone, found := k.GetHostZone(ctx, chainId)
+	if !found {
+		return errorsmod.Wrap(types.ErrHostZoneNotFound, fmt.Sprintf("Host zone %s not found", chainId))
+	}
+	if hostZone.DelegationIcaAddress == "" {
+		return errorsmod.Wrapf(types.ErrICAAccountNotFound, "no delegation account found for %s", chainId)
+	}
+
+	// Get the difference between the actual and expected validator delegations
+	valDeltaList, err := k.GetValidatorDelegationDifferences(ctx, hostZone, delegationType)
+	if err != nil {
+		return errorsmod.Wrapf(err, "unable to get validator deltas for host zone %s", chainId)
+	}
+
+	msgs, rebalacings := k.GetRebalanceICAMessages(hostZone, valDeltaList, numRebalance)
+
+	// marshall the callback
+	rebalanceCallback := types.RebalanceCallback{
+		HostZoneId:     hostZone.ChainId,
+		DelegationType: delegationType,
+		Rebalancings:   rebalacings,
+	}
+	rebalanceCallbackBz, err := k.MarshalRebalanceCallbackArgs(ctx, rebalanceCallback)
+	if err != nil {
+		return errorsmod.Wrapf(err, "unable to marshal rebalance callback args")
+	}
+
+	// Submit the rebalance ICA
+	_, err = k.SubmitTxsStrideEpoch(
+		ctx,
+		hostZone.ConnectionId,
+		msgs,
+		types.ICAAccountType_DELEGATION,
+		ICACallbackID_Rebalance,
+		rebalanceCallbackBz,
+	)
+	if err != nil {
+		return errorsmod.Wrapf(err, "Failed to SubmitTxs for %s, messages: %+v", hostZone.ChainId, msgs)
+	}
+
+	return nil
+}
+
+// Given a list of target delegation changes, builds the individual re-delegation messages by redelegating
+// from overweight validators to undersweight validators
+// Also returns the callback data for the ICA
+func (k Keeper) GetRebalanceICAMessages(
+	hostZone types.HostZone,
+	validatorDeltas []RebalanceValidatorDelegationChange,
+	numRebalance uint64,
+) (msgs []sdk.Msg, rebalancings []*types.Rebalancing) {
+	// Sort the list of delegation changes by the size of the change
+	lessFunc := func(i, j int) bool {
+		if !validatorDeltas[i].Delta.Equal(validatorDeltas[j].Delta) {
+			return validatorDeltas[i].Delta.LT(validatorDeltas[j].Delta)
+		}
+		// use name as a tie breaker if deltas are equal
+		return validatorDeltas[i].ValidatorAddress < validatorDeltas[j].ValidatorAddress
+	}
+	sort.SliceStable(validatorDeltas, lessFunc)
+
+	// QUESTION: Does anyone have a preference on the naming convention here? A couple options:
+	//  * Overweight/Underweight (current)
+	//  * Surplus/Deficit
```

Comment:
Surplus/Deficit feels more accurate than Overweight/Underweight, because the * amount * is in surplus/deficit relative to the weight. The weight it always correct, so overweight/underweight is technically not correct

Reply:
Agree

Reply:
agreed as will. Will update if @shellvish's on the same page!

Reply:
done